timesketch/lib/utils.py

```diff
@@ -83,97 +90,99 @@ def _scrub_special_tags(dict_obj):
             _ = dict_obj.pop(field)
 
 
-def read_and_validate_csv(file_handle, delimiter=','):
+def _get_unix_time(date):
+    """Convert a Pandas Timestamp object to a UNIX time in microseconds"""
+    return str(time.mktime(
+        date.to_pydatetime().utctimetuple()) * 1e6 + date.microsecond)
+
+
+def _validate_csv_fields(mandatory_fields, data):
+    """Validate parsed CSV fields against mandatory fields.
+
+    Args:
+        mandatory_fields: a list of fields that must be present.
+        data: a DataFrame built from the ingested file.
+
+    Raises:
+        RuntimeError: if there are missing fields.
+    """
+    mandatory_set = set(mandatory_fields)
+    parsed_set = set(data.columns)
+
+    if mandatory_set.issubset(parsed_set):
+        return
+
+    raise RuntimeError('Missing fields in CSV header: {0:s}'.format(
+        ','.join(list(mandatory_set.difference(parsed_set)))))
+
+
+def read_and_validate_csv(file_handle, delimiter=',',
+                          mandatory_fields=TIMESKETCH_FIELDS):
     """Generator for reading a CSV file.
 
     Args:
         file_handle: a file-like object containing the CSV content.
         delimiter: character used as a field separator, default: ','
+        mandatory_fields: list of fields that must be present in the CSV header.
 
     Raises:
         RuntimeError: when there are missing fields.
         DataIngestionError: when there are issues with the data ingestion.
     """
-    # Columns that must be present in the CSV file.
-    mandatory_fields = ['message', 'datetime', 'timestamp_desc']
-
     # Ensures delimiter is a string.
     if not isinstance(delimiter, six.text_type):
         delimiter = codecs.decode(delimiter, 'utf8')
 
-    # Due to issues with python2.
-    if six.PY2:
-        delimiter = str(delimiter)
-
-    reader = csv.DictReader(file_handle, delimiter=delimiter)
-    csv_header = reader.fieldnames
-    missing_fields = []
-    # Validate the CSV header
-    for field in mandatory_fields:
-        if field not in csv_header:
-            missing_fields.append(field)
-    if missing_fields:
-        raise RuntimeError(
-            'Missing fields in CSV header: {0:s}'.format(
-                ','.join(missing_fields)))
+    header_reader = pandas.read_csv(file_handle, sep=delimiter, nrows=0)
+    _validate_csv_fields(mandatory_fields, header_reader)
     try:
-        for row in reader:
-            # There is a condition in which the CSV reader can read a
-            # single lines as multiple, causing issues with importing.
-            # TODO: Swap the CSV library for the user of pandas.
-            if not row:
-                continue
-            if not row['datetime']:
+        reader = pandas.read_csv(file_handle, sep=delimiter,
+                                 chunksize=DEFAULT_CHUNK_SIZE)
+        for chunk in reader:
+            skipped_rows = chunk[chunk['datetime'].isnull()]
+            for row in skipped_rows.iterrows():
```

Comment:
do we want to do that for all of these rows?

or just say

```
if not skipped_rows.empty:
  logger.warning(f'{len(skipped_rows)} rows skipped since they were missing a datetime field or it was empty')
```

I'm just wondering whether we need all of them to be logged here?